### PR TITLE
[DO NOT REVIEW] Refactor Connector and SyncJob

### PR DESCRIPTION
**Pull request by wangch079**
This PR tries to refactor `connectors-python`:
1. Make `Connector` and `SyncJob` class only data representation of a doc in `.elastic-connectors`/`.elastic-connectors-sync-jobs`. It also contains some self-mutating methods, e.g. `claim` and `suspend` methods in `SyncJob`, `heartbeat` and `prepare` methods in `Connector`.
2. Make `ConnectorIndex` and `SyncJobIndex` responsible for index-level operations only, e.g. retrieve `supported_connectors` in `ConnectorIndex`, retrieve `pending_jobs` in `SyncJobIndex`.
3. Make `SyncService` only responsible for scheduling jobs, and the job execution is managed by a new class `SyncJobRunner`.

Then we realized the scope of this PR is too large, so we decided to break it down into multiple PRs:
1. #430 
2. #431 
3. #432 
4. #449
5. #440 
6. #442 
7. #443 
8. #434
9. #460 
10. #474
11. #495
12. #498
13. #491 
14. #500 

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [x] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: connectors/byoc.py [hunk 1]
```diff
@@ -164,22 +167,20 @@ def build_docs_query(self, native_service_types=None, connectors_ids=None):
...
-            self,
-            doc_source["_id"],
-            doc_source["_source"],
+            doc_source,
             bulk_options=self.bulk_options,
         )
 
 
 class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
```
Comment (wangch079) [inline connectors/byoc.py:175]: Before it's getting too complicated, I would like to get the team's opinion on this. 

Currently each entity instance (of class `Connector`/`SyncJob`) holds an instance variable `index` (instance of `ConnectorIndex`/`SyncJobIndex`, which is the child class of `ESClient`), so all API calls to Elasticsearch are through this variable).

I want to introduce some static methods, e.g. `pending_jobs` in `SyncJob`, so I can get all pending jobs by calling `SyncJob.pending_jobs()`. But this won't work as the class doesn't hold the index instance. That's why I introduced a class variable (same in class `Connector`), and they should be set in `cli.py`.

cc @elastic/ingestion-team

File: connectors/cli.py [hunk 2]
```diff
@@ -140,10 +141,16 @@ def run(args):
...
+        Connector.es_index = ConnectorIndex(config["elasticsearch"])
+        SyncJob.es_index = SyncJobIndex(config["elasticsearch"])
         return loop.run_until_complete(coro)
     except asyncio.CancelledError:
         return 0
     finally:
+        if Connector.es_index is not None:
+            loop.run_until_complete(Connector.es_index.close())
+        if SyncJob.es_index is not None:
+            loop.run_until_complete(SyncJob.es_index.close())
```
Comment (wangch079) [inline connectors/cli.py:153]: Not really sure about this. cc @tarekziade

File: connectors/byoc.py (hunk 1 above)
Comment (vidok) [reply connectors/byoc.py:175]: Should `pending_jobs ` be a part of `SyncJobIndex`? `SyncJobIndex` has the ability to query jobs and then create a list of `SyncJob` objects based on the elasticsearch documents (see [_create_object](https://github.com/elastic/connectors-python/blob/main/connectors/byoc.py#L702).

I think `SyncJobIndex` should be responsible for any sync job lookups and `SyncJob` should represent a single document.

Comment (wangch079) [reply connectors/byoc.py:175]: OK. That's different from my understanding, and the Ruby implementation.

So you are suggesting, the index class (`ConnectorIndex` and `SyncJobIndex`) should be responsible for any business logic related read/write. The entity class (`Connector` and `SyncJob`) only represent a single document, and is responsible for mutating itself, e.g. `job.claim()`, `job.cancel()`? 


Or do you think we could make it a simple data object, and move those mutating methods to the index class as well? This could simply it significantly as it doesn't have to hold an index instance.

File: connectors/byoc.py [hunk 3]
```diff
@@ -125,7 +125,10 @@ async def update_filtering_validation(
             retry_on_conflict=RETRY_ON_CONFLICT,
         )
 
-    def build_docs_query(self, native_service_types=None, connectors_ids=None):
+    @staticmethod
```
Comment (tarekziade) [inline connectors/byoc.py:128]: why a staticmethod? if it's not using any of the class variables, make this a seperate function

File: connectors/byoc.py [hunk 4]
```diff
@@ -164,22 +167,20 @@ def build_docs_query(self, native_service_types=None, connectors_ids=None):
...
-            doc_source["_id"],
-            doc_source["_source"],
+            doc_source,
             bulk_options=self.bulk_options,
         )
 
 
 class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
+    es_index = None
```
Comment (tarekziade) [inline connectors/byoc.py:176]: don't use class variables, put this in `__init__`

File: connectors/byoc.py (hunk 1 above)
Comment (vidok) [reply connectors/byoc.py:175]: > So you are suggesting, the index class (ConnectorIndex and SyncJobIndex) should be responsible for any business logic related read/write. 

Yes, logic like pulling documents and create object iterators, pagination. Everything we need to work with Elasticsech indices. 

an abstract code might look like: 

```
for job in sync_job_index.pending_jobs():
  job.claim()
  # execution
  job.finish()
```

> The entity class (Connector and SyncJob) only represent a single document, and is responsible for mutating itself, e.g. job.claim(), job.cancel()?

Yes!

> Or do you think we could make it a simple data object, and move those mutating methods to the index class as well? This could simply it significantly as it doesn't have to hold an index instance.

I don't think so. I think it's pretty good to access the ES client from the entity classes but we can choose another way to propagate ES connection. Now it's an es_index object which is not the ideal way but this can be easily improved because I encapsulated object creation.

Comment (tarekziade) [reply connectors/byoc.py:175]: in your refactorings, make sure you initialize stuff in `__init__` and not add class variables (as opposed to object variables) -- there are very few cases where we want this.
using class variables means you share the same object across all instances, this adds edge issues

File: connectors/byoc.py [hunk 5]
```diff
@@ -232,10 +233,10 @@ async def done(self, indexed_count=0, deleted_count=0, exception=None):
 
         job_def["status"] = e2str(self.status)
 
-        return await self.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
+        return await self.es_index.client.update(index=JOBS_INDEX, id=self.job_id, doc=job_def)
 
-    @classmethod
-    def transform_filtering(cls, filtering):
+    @staticmethod
```
Comment (tarekziade) [inline connectors/byoc.py:238]: don't use static methods

File: connectors/byoc.py [hunk 6]
```diff
@@ -707,12 +705,12 @@ def _create_object(self, doc_source):
             SyncJob
         """
         return SyncJob(
-            self,
             connector_id=doc_source["_source"]["connector"]["id"],
             doc_source=doc_source["_source"],
         )
 
-    def pending_job_query(self, connectors_ids):
+    @staticmethod
```
Comment (tarekziade) [inline connectors/byoc.py:712]: same remark

**Review by tarekziade — COMMENTED** [review]
This looks good for a first draft, awesome. 

About `staticmethod` vs `classmethod` vs regular methods: you should always use regular methods -- when you use static or class methods you are using classes as a namespace, not OOP elements anymore. There are a few spots in filtering rules where it's ok, but here I don't see a use case

File: connectors/cli.py [hunk 7]
```diff
@@ -140,10 +141,16 @@ def run(args):
     coro = _start_service(config, args, loop)
 
     try:
+        Connector.es_index = ConnectorIndex(config["elasticsearch"])
```
Comment (tarekziade) [inline connectors/cli.py:144]: There's a small design issue. You want to create your indexes and pass them to the Services you create, and deal with their shutdown there.  The `cli` module should only deal with `Service` classes.

File: connectors/byoc.py (hunk 1 above)
Comment (wangch079) [reply connectors/byoc.py:175]: Thanks for the comments.

Seems the team is not in favor of class variables. I will make the change accordingly.

Comment (tarekziade) [reply connectors/byoc.py:175]: yeah class variables break the class contract because it becomes a namespace and loses its OOP behaviors. It's very helpful in met-programming but in other cases, not. In Python it's completely fine also to use plain functions when a method would be static. It's cleaner

File: connectors/services/sync.py [hunk 8]
```diff
@@ -64,32 +71,61 @@ async def _one_sync(self, connector, es, sync_now):
...
+                return
+
+            if connector.features.sync_rules_enabled():
+                await validate_filtering(
+                    connector, self.connectorIndex, ValidationTarget.DRAFT
+                )
+
+            if not sync_now:
+                next_sync = connector.next_sync()
+                if next_sync == SYNC_DISABLED or next_sync - self.idling > 0:
```
Comment (vidok) [inline connectors/services/sync.py:91]: 2 things here: 

* Why do need the second condition against `SYNC_DISABLED`?
* I would move this logic (the entire block from L89 to L100  into a separate method

File: connectors/source.py [hunk 9]
```diff
@@ -235,3 +235,11 @@ def get_data_sources(config):
     """Returns an iterator of all registered sources."""
     for name, fqn in config["sources"].items():
         yield get_source_klass(fqn)
+
+
+def get_data_source_dict(config):
```
Comment (vidok) [inline connectors/source.py:240]: I would not use data type in the functions` names. Does it technically return data source configuration?

File: connectors/sync_job_runner.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            return
+
+        try:
```
Comment (vidok) [inline connectors/sync_job_runner.py:53]: It's super hard to read such a long block of code. Is there a way to split it into several ones?

File: connectors/byoc.py [hunk 11]
```diff
@@ -81,31 +67,53 @@ class DataSourceError(Exception):
...
-            index=CONNECTORS_INDEX,
-            id=connector.id,
-            doc=document,
-        )
+
+    async def heartbeat(self, doc_id):
+        doc = {
+            "last_seen": iso_utc(),
+        }
+        await self.update(doc_id, doc)
```
Comment (tarekziade) [inline connectors/byoc.py:116]: one line works for such a simple change:

```
await self.update(doc_id, {"last_seen": iso_utc()})
```

File: connectors/byoc.py [hunk 12]
```diff
@@ -125,114 +133,144 @@ async def update_filtering_validation(
...
             doc_source = dict()
+        self.id = doc_source.get("_id")
+        self._source = doc_source.get("_source", {})
+
+    def get(self, *keys, default=None):
+        value = self._source
+        for key in keys:
+            if not isinstance(value, dict):
+                return default
+            value = value.get(key)
```
Comment (tarekziade) [inline connectors/byoc.py:166]: if you have several keys, `value` is overridden on the second round of this loop. is this intended?

Comment (tarekziade) [reply connectors/byoc.py:166]: Overriding the variable `value` also means you won't have `self._source` in it.

File: connectors/byoc.py [hunk 13]
```diff
@@ -278,6 +316,8 @@ def get_filter(self, filter_state="active", domain=DEFAULT_DOMAIN):
 
 class PipelineSettings:
     def __init__(self, pipeline):
```
Comment (tarekziade) [inline connectors/byoc.py:318]: `pipeline=None` ?

File: connectors/byoc.py [hunk 14]
```diff
@@ -347,131 +387,52 @@ def nested_get(dictionary, keys_, default_=None):
...
-    @status.setter
-    def status(self, value):
-        if isinstance(value, str):
-            value = Status[value.upper()]
-        if not isinstance(value, Status):
-            raise TypeError(value)
+    @property
+    def last_seen(self):
+        last_seen = self.get("last_seen")
+        logger.info(f"last_seen = {last_seen}, type = {last_seen.__class__}")
```
Comment (tarekziade) [inline connectors/byoc.py:402]: shouldn't this be `debug` ?

File: connectors/byoc.py [hunk 15]
```diff
@@ -481,211 +442,168 @@ def next_sync(self):
...
-
-        self.sync_now = self.doc_source["sync_now"] = False
-        self.doc_source["last_sync_status"] = e2str(job.status)
-        self.status = Status.CONNECTED
-        await self.sync_doc()
+    async def reset_sync_now_flag(self):
+        doc = {
+            "sync_now": False,
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
```
Comment (tarekziade) [inline connectors/byoc.py:454]: can be a one liner

File: connectors/byoc.py [hunk 16]
```diff
@@ -481,211 +442,168 @@ def next_sync(self):
...
+            "status": e2str(Status.ERROR),
+            "error": str(error),
+        }
+        await self.index.update(doc_id=self.id, doc=doc)
+
+    async def sync_done(self, job):
+        job_status = JobStatus.ERROR if job is None else job.status
+        job_error = "Couldn't find the job" if job is None else job.error
+        if job_error is None and job_status == JobStatus.ERROR:
+            job_error = "unknown error"
```
Comment (tarekziade) [inline connectors/byoc.py:474]: let's make this a CONST at the top of the file

File: connectors/services/sync.py [hunk 17]
```diff
@@ -64,32 +71,61 @@ async def _one_sync(self, connector, es, sync_now):
...
         except InvalidFilteringError as e:
             logger.error(e)
-            return
-        finally:
-            await connector.close()
 
     async def _run(self):
         """Main event loop."""
-        self.connectors = ConnectorIndex(self.es_config)
+        self.connectorIndex = ConnectorIndex(self.es_config)
```
Comment (tarekziade) [inline connectors/services/sync.py:127]: use PEP 8

```
self.connector_index
self.sync_job_index
```

**Review by tarekziade — COMMENTED** [review]
neat, lots of cool stuff

File: connectors/es/index.py [hunk 18]
```diff
@@ -40,7 +40,32 @@ def _create_object(self, doc):
...
+        return self._create_object(resp)
+
+    async def index(self, doc):
+        resp = await self.client.index(index=self.index_name, document=doc)
+        return resp["_id"]
+
+    async def update(self, doc_id, doc):
+        await self.client.update(index=self.index_name, id=doc_id, doc=doc)
+
+    async def document_count(self):
```
[the remark this answers is not in the document]
Comment (wangch079) [reply connectors/es/index.py:64]: All the stats fields in `.elastic-connectors-sync-jobs` are singular, e.g. `indexed_document_count`, `total_document_count`. Just want to be consistent.

File: connectors/byoc.py (hunk 13 above)
Comment (wangch079) [reply connectors/byoc.py:318]: To instantiate an instance of `PipelineSettings`, it's expected to provide a `pipeline`, so I am inclined to make it mandatory. But we couldn't stop someone using `pipeline = PipelineSettings(pipeline=None)`, therefore I added additional protection:
```python
if pipeline is None:
    pipeline = {}
```

File: connectors/byoc.py (hunk 14 above)
Comment (wangch079) [reply connectors/byoc.py:402]: This log shouldn't be here. I added this for my own testing,  because I'm not sure whether the `last_seen` will be parsed as string or time in the response json. I will remove it eventually.

File: connectors/services/sync.py (hunk 8 above)
Comment (wangch079) [reply connectors/services/sync.py:91]: I think the inner if-else branch is to log different messages. I just copied this block from somewhere else. 

Will furthur refactor it.

File: connectors/source.py (hunk 9 above)
Comment (wangch079) [reply connectors/source.py:240]: No, it returns a service type - source class dictionary. I just followed the pattern of the above method `get_data_sources`, and now I reallized it was named wrongly. Probably we should rename the above method to `get_source_klasses` and this method to `get_source_klass_dict`

File: connectors/byoc.py (hunk 12 above)
Comment (wangch079) [reply connectors/byoc.py:166]: Yes, it's intended. 

> Overriding the variable `value` also means you won't have `self._source` in it.

Yes, we only need `value` pointing to `self._source` in the beginning, then we want to dig with `keys` provided, it's like the [dig method in Ruby Hash](https://ruby-doc.org/core-3.1.1/Hash.html#method-i-dig)

Comment (tarekziade) [reply connectors/byoc.py:166]: ok I see 👍

File: connectors/byoc.py (hunk 13 above)
Comment (tarekziade) [reply connectors/byoc.py:318]: yes but this protection is just for None. This would break too:

```
pipeline = PipelineSettings(pipeline=1)
```

If pipeline is mandatory, you have to trust it's a pipeline. If there's no way to trust it, you can do:

```
if not isinstance(pipeline, dict):
    raise TypeError(pipeline)
```

but usually, we just trust it -- and you will get an error on the first attempt to use a dict method ("if it quacks like a duck, it's a duck")

File: connectors/source.py (hunk 9 above)
Comment (vidok) [reply connectors/source.py:240]: What is `source class dictionary`?

File: connectors/es/index.py [hunk 19]
```diff
@@ -40,7 +40,32 @@ def _create_object(self, doc):
         """
         raise NotImplementedError
 
-    async def get_all_docs(self, query=None, page_size=DEFAULT_PAGE_SIZE):
+    async def fetch_by_id(self, doc_id):
+        await self.client.indices.refresh(index=self.index_name)
```
Comment (artem-shelkovnikov) [inline connectors/es/index.py:44]: I'm curious about it - should we always call it or should we never call it?

I _think_ we do refresh stuff automatically, but also awaiting a `refresh` does not guarantee that the data is refreshed, only that the `refresh` operation was issued and accepted, is it correct?

File: connectors/es/index.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply connectors/es/index.py:64]: I also think it's singular - think of phrase like "word count" or "head count". I also googled a bit and saw similar examples, but haven't been able to find the actual rule:

https://english.stackexchange.com/questions/349475/singular-or-plural-noun-preceding-count
https://www.usingenglish.com/forum/threads/plural-or-singular-when-using-with-the-word-count.242655/

File: connectors/services/sync.py (hunk 8 above)
Comment (artem-shelkovnikov) [reply connectors/services/sync.py:91]: It can be a big refactoring overall (given we want to test it fully, I think there are missing tests for scheduling), so can probably be done in a separate PR

File: connectors/source.py (hunk 9 above)
Comment (artem-shelkovnikov) [reply connectors/source.py:240]: I see it's just a "string" -> "string" dictionary, key is service_type, value is class FQN (see config.yml's "sources" section: https://github.com/elastic/connectors-python/blob/main/config.yml#L38)

File: connectors/sync_job_runner.py [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+            await self.connector.sync_done(self.sync_job)
+
+            logger.info(
+                f"Sync done: {indexed_count} indexed, {doc_deleted} "
+                f" deleted. ({int(time.time() - self._start_time)} seconds)"
+            )
+
+    async def _claim_job(self):
+        try:
+            await self.sync_job.claim()
```
Comment (artem-shelkovnikov) [inline connectors/sync_job_runner.py:136]: Hmmm, interesting, so with this PR our connectors will be more multi-tenant friendly?

File: connectors/byoc.py (hunk 13 above)
Comment (wangch079) [reply connectors/byoc.py:345]: For this case, I think we do need `pipeline` in the `__init__` method. so I think we can keep it as 
```python
def __init__(self, pipeline):
```

Do you think it's unnecessary to have this?
```python
if pipeline is None:
    pipeline = {}
```

File: connectors/source.py (hunk 9 above)
Comment (wangch079) [reply connectors/source.py]: It's not a `string` (service type) -> `string` (FQN) dictionary. It's a `string` (service type) -> `class` (child classes of `BaseDataSource`). 

It serve exactly the same purpose as the [REGISTRY](https://github.com/elastic/connectors-ruby/blob/9d5be7a79ee18c794403d14b1dcfc2c574786771/lib/connectors/registry.rb#L40) of `connectors-ruby`

File: connectors/es/index.py (hunk 19 above)
Comment (wangch079) [reply connectors/es/index.py:44]: The `refresh` call is synchronous, so it's guaranteed that we get refreshed data.

For most of the use cases (if not all), we use `search`. When we use `fetch_by_id`, it normally means we need to refresh the document, because there will likely be an update happened. That's why we need the refresh.

Yes, ES does auto-refresh every second by default, and the official guide suggests not to use the refresh API, but [use `refresh=wait_for` when updating](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-refresh.html#refresh-api-desc).

[`refresh=wait_for`](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html) will wait for a periodic refresh to happen before returning. If we apply this option to our update request, it may take, the worst case 1 second, to get the response. cc @tarekziade

File: connectors/sync_job_runner.py (hunk 20 above)
Comment (wangch079) [reply connectors/sync_job_runner.py:135]: But we don't have concurrency control at all.

File: connectors/source.py [hunk 21]
```diff
@@ -103,6 +105,9 @@ def get_fields(self):
     def is_empty(self):
         return len(self._config) == 0
 
+    def to_dict(self):
+        return deepcopy(self._raw_config)
```
Comment (tarekziade) [inline connectors/source.py:109]: `deepcopy` is overkill, even `copy` as long as the config is mappings of mappings -- you can use `dict`

```
>>> o = {'a': {'b': 'c'}}
>>> dict(o)
{'a': {'b': 'c'}}
```

File: connectors/source.py [hunk 22]
```diff
@@ -272,7 +277,15 @@ def get_source_klass(fqn):
     return getattr(module, klass_name)
 
 
-def get_data_sources(config):
+def get_source_klasses(config):
```
Comment (tarekziade) [inline connectors/source.py:280]: better name!

File: connectors/source.py [hunk 23]
```diff
@@ -272,7 +277,15 @@ def get_source_klass(fqn):
...
     """Returns an iterator of all registered sources."""
     for name, fqn in config["sources"].items():
         yield get_source_klass(fqn)
+
+
+def get_source_klass_dict(config):
```
Comment (tarekziade) [inline connectors/source.py:286]: makes me wonder if we don't want to add a `name` property to the source klass...

File: connectors/sync_job_runner.py [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+            if not await data_provider.changed():
+                logger.debug(
+                    f"No change in {self.sync_job.service_type} data provider, skipping..."
+                )
+                return
+
+            logger.debug(f"Syncing '{self.sync_job.service_type}'")
+            logger.debug(f"Pinging the {self.source_klass} backend")
+            await data_provider.ping()
+            await asyncio.sleep(0)
```
Comment (tarekziade) [inline connectors/sync_job_runner.py:67]: you can remove this one because you have another await right after.

```
await asyncio.sleep(0)
```

is useful in a block of sync calls that loops, to return execution to the event loop, and that happens on any await call

File: connectors/sync_job_runner.py [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+
+            mappings = Mappings.default_text_fields_mappings(
+                is_connectors_index=True,
+            )
+
+            logger.debug("Preparing the content index")
+            await self.elastic_server.prepare_content_index(
+                self.sync_job.index_name, mappings=mappings
+            )
+            await asyncio.sleep(0)
```
Comment (tarekziade) [inline connectors/sync_job_runner.py:77]: same remark

File: connectors/source.py (hunk 23 above)
Comment (wangch079) [reply connectors/source.py:286]: We should. we actually have `name` property in Ruby.

Currently the `list` command uses `__doc__` of each source klass:

https://github.com/elastic/connectors-python/blob/c1baa3a0bdd74a173bedb545dad01df276e3faef/connectors/cli.py#L135


I can make the change in a separate PR.

But even we have `name` and `service_type` in each source klass, we still need a dict to quickly loop up the source klass

Comment (wangch079) [reply connectors/source.py:286]: PR is here: https://github.com/elastic/connectors-python/pull/413

File: connectors/byoc.py [hunk 26]
```diff
@@ -81,180 +70,209 @@ class DataSourceError(Exception):
...
+        return self.get("total_document_count", default=0)
+
+    async def claim(self):
+        doc = {
+            "status": e2str(JobStatus.IN_PROGRESS),
+            "started_at": iso_utc(),
+            "last_seen": iso_utc(),
+            "worker_hostname": socket.gethostname(),
         }
+        await self.index.update(doc_id=self.id, doc=doc)
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:244]: No concurrency handling here yet, right?

File: connectors/byoc.py [hunk 27]
```diff
@@ -297,6 +315,9 @@ def get_filter(self, filter_state="active", domain=DEFAULT_DOMAIN):
             Filter(),
         )
 
+    def to_list(self):
+        return list(self.filtering)
+
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:320]: Looks redundant, maybe better to get rid of it?

File: connectors/byoc.py [hunk 28]
```diff
@@ -320,17 +341,35 @@ def has_advanced_rules(self):
...
 
     def __repr__(self):
         return (
             f"Pipeline {self.name} <binary: {self.extract_binary_content}, "
             f"whitespace {self.reduce_whitespace}, ml inference {self.run_ml_inference}>"
         )
 
+    def to_dict(self):
+        return dict(self._pipeline)
+
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:372]: I think it can be replaced with `__dict__` function and calling `dict` on instance of PipelineSettings should call it. cc @tarekziade is it correct?

File: connectors/byoc.py (hunk 27 above)
Comment (artem-shelkovnikov) [reply connectors/byoc.py:320]: Ah, it's probably better to implement Python's Abstract Base Class that allows certain operations over filtering - maybe just `__iter__` and `__get_item__` and `__len__` will be sufficient - see https://docs.python.org/3/library/collections.abc.html

File: connectors/byoc.py [hunk 29]
```diff
@@ -529,217 +501,110 @@ def next_sync(self):
         if self.sync_now:
             logger.debug("sync_now is true, syncing!")
             return 0
-        if not self.scheduling["enabled"]:
+        if not self.scheduling.get("enabled", False):
```
Comment (artem-shelkovnikov) [inline connectors/byoc.py:504]: Nice!

File: connectors/services/sync.py [hunk 30]
```diff
@@ -40,63 +45,68 @@ def __init__(self, config, args):
...
-            await connector.error(e)
-            logger.critical(e, exc_info=True)
-            self.raise_if_spurious(e)
+        if not await self._prepare_connector(connector):
             return
 
         try:
             # the heartbeat is always triggered
-            connector.start_heartbeat(self.hb)
+            await connector.heartbeat(self.heartbeat_interval)
```
Comment (artem-shelkovnikov) [inline connectors/services/sync.py:66]: I'd probably leave `heartbeat_interval` out of the function and do the condition here, e.g:

```python
if self.should_trigger_heartbeat():
    connector.heartbeat()
```

Comment (artem-shelkovnikov) [reply connectors/services/sync.py:66]: But it's more of a nit fixable later

**Review by artem-shelkovnikov — APPROVED** [review]
I'm hesitant about the change purely because of its size - I intuitively understand that it extracts running sync jobs to a separate class, but also a lot of small side-changes are making it harder to understand in the sense of whether new problems are introduced or not.

I'll carefully say that if we want to merge this change, we should do it ASAP - it's been around for 2 weeks.

So unless somebody has blocking concerns (cc @tarekziade), let's actually land it asap and fix bugs on the way.

File: connectors/byoc.py (hunk 26 above)
Comment (wangch079) [reply connectors/byoc.py:244]: No, do we have concurrency control in `connectors-python`? I think it will be handled separately?

File: connectors/byoc.py (hunk 27 above)
Comment (wangch079) [reply connectors/byoc.py:320]: The reason why I added `to_list` to class `Filtering`, is because it was used to update the connector, where a list is required instead of an object:

https://github.com/elastic/connectors-python/blob/59d90d42ba9c01163c0455227d3637ef805d7028/connectors/byoc.py#L118-L134

Similar reason for `to_dict` of class `PipelineSettings`:

https://github.com/elastic/connectors-python/blob/59d90d42ba9c01163c0455227d3637ef805d7028/connectors/byoc.py#L649

I don't intend to make `Filtering` or `PipelineSettings` a collection.

File: connectors/byoc.py (hunk 26 above)
Comment (artem-shelkovnikov) [reply connectors/byoc.py:244]: Let's do separately, yes. Just checking that you're not addressing it here.

File: connectors/byoc.py (hunk 28 above)
Comment (wangch079) [reply connectors/byoc.py:372]: I don't think `__dict__` is for this purpose ([ref](https://docs.python.org/3/library/stdtypes.html#object.__dict__)):
>A dictionary or other mapping object used to store an object’s (writable) attributes.

Comment (artem-shelkovnikov) [reply connectors/byoc.py:372]: Sorry my bad. What I've meant is that if it's used as hash, then the class might actually expose dictionary functions so that it could be done outside of the object, e.g:

pipeline_dict = dict(Pipeline) or something else.

In short term this code works, but I believe we can improve on it later - in later PRs.

**Comment by wangch079** [comment]
>I'm hesitant about the change purely because of its size

I agree. I'm not a fan of huge PR either. I tried to keep it small when I worked on this PR but it kept growing and seemed out of control. 

What I have done for this PR:
1. Tested locally for both custom connector and native connector.
2. Added tests and make sure we have > 92% test coverage.

>let's actually land it asap and fix bugs on the way.

I'd say yes. Having this merged will speed up the whole team, not only myself. With the above 2 points, it won't break `main`. There may still be bugs, which is more easier to spot if the whole team starts working on `main` branch after it's merged, instead of only me.

**Comment by wangch079** [comment]
Hi @artem-shelkovnikov , @tarekziade , I implemented the changes of https://github.com/elastic/connectors-python/pull/376 in this commit https://github.com/elastic/connectors-python/pull/375/commits/b306deca18e514e35fe39dd21b7f23b4343df46f.

When I try to shut down the service during job execution, I can see it reached this line:

https://github.com/elastic/connectors-python/blob/b306deca18e514e35fe39dd21b7f23b4343df46f/connectors/sync_job_runner.py#L97-L98,

and it also reached the `finally` block:

https://github.com/elastic/connectors-python/blob/b306deca18e514e35fe39dd21b7f23b4343df46f/connectors/sync_job_runner.py#L103

but it didn't proceed further, so the job stays at `in_progress` status. it seems it won't wait for the `finally` block to complete.

Do you spot anything wrong in my implementation? Else I will continue tomorrow.

**Review by tarekziade — COMMENTED** [review]
Following up on the discussion in https://elastic.slack.com/archives/C01795T48LQ/p1675759094917429

The patch is doing a lot of changes in the core that could probably happen in smaller patches, and at this point, I think it's a bit risky to land as-is for 8.7 FF.

I would advise explaining in the issue the goal of this refactoring and if possible splitting some of the work in some other PRs. I've identified several changes that can land with no risks in smaller PRs.

I would also want to see a couple of e2e test runs, with the memory and speed metrics before and after.

**Comment by tarekziade** [comment]
>  let's actually land it asap and fix bugs on the way.

sorry, this is not acceptable. we don't break main and then fix it

**Comment by tarekziade** [comment]
> > let's actually land it asap and fix bugs on the way.
> 
> I'd say yes. Having this merged will speed up the whole team, not only myself. With the above 2 points, it won't break `main`. There may still be bugs, which is more easier to spot if the whole team starts working on `main` branch after it's merged, instead of only me.

I don't understand. If your PR has bugs and you know about it, and if it does not "break main" it means you miss some tests.
and you can't push that on the whole team. It's the dev responsibility to provide a patch that is tested with no known bugs and tested

**Comment by wangch079** [comment]
>I don't understand. If your PR has bugs and you know about it

What I meant here is that I have tested it locally and it worked, and I have added all the tests. But there are still possibly some bugs that I didn't spot.

Therefore:
>There may still be bugs, which is more easier to spot if the whole team starts working on main branch after it's merged, instead of only me.

**Review by tarekziade — CHANGES_REQUESTED** [review]
see previous comment

**Comment by tarekziade** [comment]
> > There may still be bugs, which is more easier to spot if the whole team starts working on main branch after it's merged, instead of only me.

Can you elaborate in the issue of what was the goal of this patch? there's do description. 
I am looking for a problem, how you fix it and a list of changes to do so.

The problem I see is that you added unrelated changes in the past few days that were not initially planned, growing the patch to a bigger one, so I'd like to understand what was the initial intent in this refactoring to try to minimize the risks of landing it.

**Comment by tarekziade** [comment]
nightly build for this branch https://buildkite.com/elastic/connectors-python-nightly/builds/302

**Comment by tarekziade** [comment]
the patch break the connectors in the e2e test and also blocks on the cleanup service.

Screenshot 2023-02-07 at 20 22 31

**Comment by tarekziade** [comment]
to reproduce locally

```
make ftest NAME=mysql
```

**Comment by wangch079** [comment]
> the patch break the connectors in the e2e test and also blocks on the cleanup service.

That is because the `last_seen` is incorrectly formatted:

https://github.com/elastic/connectors-python/blob/8027897416cd3b81182cb1e3cef2c70e14eedc8c/connectors/kibana.py#L155

Will introduce a separate PR to fix it. Will also try to break this PR to multiple PRs as small as possible.

**Comment by tarekziade** [comment]
> Will also try to break this PR to multiple PRs as small as possible.

Thanks, happy to help identify them, once we have a description of the "MVP" here (Minimum Viable Patch :) )

**Comment by tarekziade** [comment]
> That is because the `last_seen` is incorrectly formatted:

but main is not broken for this e2e test, so it has something to do with this patch I would assume

**Comment by wangch079** [comment]
>but main is not broken for this e2e test, so it has something to do with this patch I would assume

Yes, here:

https://github.com/elastic/connectors-python/blob/d19c7d923438a6550936e584c8557d9d5820f3a4/connectors/byoc.py#L447-L452

File: connectors/filtering/validation.py [hunk 31]
```diff
@@ -18,19 +18,19 @@ class ValidationTarget(Enum):
 
 
 async def validate_filtering(
-    connector, index, validation_target=ValidationTarget.ACTIVE
+    connector, source_klass, validation_target=ValidationTarget.ACTIVE
```
Comment (artem-shelkovnikov) [inline connectors/filtering/validation.py:21]: This change (whole file change) can probably be in a separate PR.

File: connectors/services/job_cleanup.py [hunk 32]
```diff
@@ -65,7 +65,7 @@ async def _process_orphaned_jobs(self):
                     and job.index_name not in existing_content_indices
                 ):
                     content_indices.add(job.index_name)
-                job_ids.append(job.job_id)
+                job_ids.append(job.id)
```
Comment (artem-shelkovnikov) [inline connectors/services/job_cleanup.py:68]: This rename can be in a separate PR

Comment (artem-shelkovnikov) [reply connectors/services/job_cleanup.py:68]: Probably whole file can be updated in a separate PR

File: connectors/services/sync.py [hunk 33]
```diff
@@ -150,8 +161,50 @@ async def _run(self):
...
+            if self.connector_index is not None:
+                self.connector_index.stop_waiting()
+                await self.connector_index.close()
+            if self.sync_job_index is not None:
+                self.sync_job_index.stop_waiting()
+                await self.sync_job_index.close()
             await es.close()
         return 0
+
+    async def _prepare_connector(self, connector):
```
Comment (artem-shelkovnikov) [inline connectors/services/sync.py:173]: This change and all below can probably be in a separate PR

**Comment by tarekziade** [comment]
> Hi @artem-shelkovnikov , sorry I forgot to tag this PR as DO NOT REVIEW.

maybe you could convert it to draft ?

**Comment by wangch079** [comment]
Close this PR as it's covered by

>1. #430 
>2. #431 
>3. #432 
>4. #449
>5. #440 
>6. #442 
>7. #443 
>8. #434
>9. #460 
>10. #474
>11. #495
>12. #498
>13. #491 
>14. #500
